src/lib/components/NglMoleculeViewer.react.js

```diff
@@ -0,0 +1,274 @@
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import {Stage, Selection} from 'ngl';
+
+/**
+ * The NglMoleculeViewer is used to render schematic diagrams
+ * of biomolecules in ribbon-structure representations.
+ * Read more about the component here:
+ * https://github.com/IvoLeist/dash_ngl
+ * Read more about the used WebGL protein viewer here:
+ * https://github.com/arose/ngl
+ */
+export default class NglMoleculeViewer extends Component {
+    constructor(props) {
+        super(props);
+        this.state = {stage: null, structuresList: []};
+    }
+
+    componentDidMount() {
+        const {id, stageParameters, viewportStyle} = this.props;
+        const params = {...stageParameters};
+        const stage = new Stage(id, params);
+        stage.setSize(viewportStyle.width, viewportStyle.height);
+        this.setState({stage});
+    }
+
+    shouldComponentUpdate(prevProps, nextProps) {
+        const {stageParameters, data} = this.props;
+
+        // check if data has changed
+        if ((data !== null) & (prevProps.data !== null)) {
+            // wait for the first pdb selection after startup
+            if (nextProps.data !== undefined) {
+                return true;
+            }
+
+            // check if pdb selection has changed
+            const oldSelection = prevProps.data[0].selectedValue;
+            const newSelection = data[0].selectedValue;
+            if (oldSelection !== newSelection) {
+                return true;
+            }
+        }
+
+        // check for stage params changed
+        const oldStage = prevProps.stageParameters;
+        const newStage = stageParameters;
+
+        // save it as a helper function in an extra script?
+        const isEqual = (obj1, obj2) => {
+            const obj1Keys = Object.keys(obj1);
+            const obj2Keys = Object.keys(obj2);
+
+            if (obj1Keys.length !== obj2Keys.length) {
+                return false;
+            }
+
+            for (const objKey of obj1Keys) {
+                if (obj1[objKey] !== obj2[objKey]) {
+                    return false;
+                }
+            }
+            return true;
+        };
+
+        if (isEqual(oldStage, newStage) === false) {
+            return true;
+        }
+
+        // no update since neither the data nor the stage paramas have changed
+        return false;
+    }
+
+    componentDidUpdate() {
+        const {data, stageParameters} = this.props;
+        const {stage, structuresList} = this.state;
+
+        stage.setParameters(stageParameters);
```

Comment:
`componentDidUpdate` is called after the component renders -- why do we need to set the stage parameters here again?

Reply:
This is needed for changing the viewer style (background, render quality etc.)

Reply:
Seems like `stage` and `structuresList` could be instance variables or the result of a memoized/calculated function call that would get mutated during `render` instead of using `didUpdate` / `didMount` to mutate them (which goes against React's paradigm: states/props are immutable)

Due to the experimental nature of this component addition, I'm fine with the way this is implemented for now but we should have plan to follow up / clean up later as required.

Reply:
@Marc-Andre-Rivet I was not aware that this is against  Reacts paradigm. Thank you for pointing that out. Do you know some good resources where I can learn about the correct usage of instance variables/memoized/calculated functions ?